common/models/checkpoint.js

```diff
@@ -27,43 +28,59 @@ module.exports = function(Checkpoint) {
    * Get the current checkpoint id
    * @callback {Function} callback
    * @param {Error} err
-   * @param {Number} checkpointId The current checkpoint id
+   * @param {Number} checkpoint The current checkpoint seq
    */
-
   Checkpoint.current = function(cb) {
     var Checkpoint = this;
-    this.find({
-      limit: 1,
-      order: 'seq DESC'
-    }, function(err, checkpoints) {
-      if (err) return cb(err);
-      var checkpoint = checkpoints[0];
-      if (checkpoint) {
-        cb(null, checkpoint.seq);
-      } else {
-        Checkpoint.create({ seq: 1 }, function(err, checkpoint) {
-          if (err) return cb(err);
-          cb(null, checkpoint.seq);
-        });
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (cp) {
+          cb(null, cp.seq);
+        } else {
+          Checkpoint.init(function(err, data) {
```

Comment:
>I am little bit afraid that this can create a race condition

Oh you are right! It is prone to race condition!

>The memory connector does not implement atomic findOrCreate now, we will need to fix that as part of this work too. 

I just looked at the implementation of `upsert` (`updateOrCreate`) in memory connector and it seems there is a possibility of race condition there as well  please see [here](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/connectors/memory.js#L224)

I just wanted to know your opinion about using lock via Redis, Semaphore or... to cope with these race conditions... Thought?